### PR TITLE
rustPlatform.importCargoLock: support extra registries

### PR DESCRIPTION
**Pull request by lbpdt**
This is useful to teach `importCargoLock` how to download crates from a registry other than crates.io. Specifically, we publish our own crates to an internal registry and this feature lets us pull from it seamlessly.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/rust/import-cargo-lock.nix [hunk 1]
```diff
@@ -89,18 +93,22 @@ let
     '';
     fetchurl {
       name = "crate-${pkg.name}-${pkg.version}.tar.gz";
-      url = "https://crates.io/api/v1/crates/${pkg.name}/${pkg.version}/download";
+      url = "${registryUrl}/v1/crates/${pkg.name}/${pkg.version}/download";
       sha256 = checksum;
     };
 
+  registries = {
+    "registry+https://github.com/rust-lang/crates.io-index" = "https://crates.io/api";
```
Comment (figsoda) [inline pkgs/build-support/rust/import-cargo-lock.nix:101]: should we limit all the extra registries to be "registry+"? are there legit use cases where the source wouldn't start with "registry+"?

```suggestion
    "https://github.com/rust-lang/crates.io-index" = "https://crates.io/api";
```

File: pkgs/build-support/rust/import-cargo-lock.nix [hunk 2]
```diff
@@ -89,18 +93,22 @@ let
...
   # Fetch and unpack a crate.
   mkCrate = pkg:
     let
       gitParts = parseGit pkg.source;
     in
-      if pkg.source == "registry+https://github.com/rust-lang/crates.io-index" then
+      if builtins.hasAttr pkg.source registries then
       let
-        crateTarball = fetchCrate pkg;
+        crateTarball = fetchCrate (builtins.getAttr pkg.source registries) pkg;
```
Comment (figsoda) [inline pkgs/build-support/rust/import-cargo-lock.nix:111]: this is more readable to me

```suggestion
        crateTarball = fetchCrate registries.${pkg.source} pkg;
```

**Review by figsoda — COMMENTED** [review]
please target staging because of the amount of rebuilds

File: pkgs/build-support/rust/import-cargo-lock.nix (hunk 1 above)
Comment (lbpdt) [reply pkgs/build-support/rust/import-cargo-lock.nix:101]: As far as I know, all other use cases start with `git+` and they are already covered by other parameters to this function. Happy to chop off the `registry+` prefix in this API.

File: pkgs/build-support/rust/import-cargo-lock.nix [hunk 3]
```diff
@@ -10,6 +10,10 @@
   # Allow `builtins.fetchGit` to be used to not require hashes for git dependencies
 , allowBuiltinFetchGit ? false
 
+  # Additional registries to pull sources from
+  #   { "https://<API URL>" = "https://<download URL>"; }
```
Comment (figsoda) [inline pkgs/build-support/rust/import-cargo-lock.nix:14]:
```suggestion
  #   { "https://<registry index URL>" = "https://<download URL>"; }
```

**Review by figsoda — APPROVED** [review]
otherwise lgtm

File: pkgs/build-support/rust/import-cargo-lock.nix (hunk 3 above)
Comment (lbpdt) [reply pkgs/build-support/rust/import-cargo-lock.nix:14]: Updated

**Review by figsoda — APPROVED** [review]

**Comment by lbpdt** [comment]
Thanks for the review, @figsoda

**Review by bhipple — APPROVED** [review]
LGTM!

File: pkgs/build-support/rust/import-cargo-lock.nix [hunk 4]
```diff
@@ -89,18 +93,23 @@ let
     '';
     fetchurl {
       name = "crate-${pkg.name}-${pkg.version}.tar.gz";
-      url = "https://crates.io/api/v1/crates/${pkg.name}/${pkg.version}/download";
+      url = "${downloadUrl}/v1/crates/${pkg.name}/${pkg.version}/download";
       sha256 = checksum;
     };
 
+  registries = {
+    "https://github.com/rust-lang/crates.io-index" = "https://crates.io/api";
```
Comment (figsoda) [inline pkgs/build-support/rust/import-cargo-lock.nix:101]: I thought about it again, maybe we should use `https://crates.io/api/v1/crates` instead, as documented [here](https://doc.rust-lang.org/cargo/reference/registry-index.html#index-configuration). What do you think?

Comment (lbpdt) [reply pkgs/build-support/rust/import-cargo-lock.nix:101]: Probably more consistent indeed, I'll send a follow-up

Comment (lbpdt) [reply pkgs/build-support/rust/import-cargo-lock.nix:101]: #222238

**Pull request merged**